src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,558 @@
+                        cf_key_count += 1;
+                        cf_size += l;
+                        if let Some(ref limiter) = io_limiter {
+                            if bytes >= base {
+                                bytes = 0;
+                                limiter.request(base);
+                            }
+                            bytes += l as i64;
+                        }
+                        writer.put(key, value)?;
+                        stale = stale_for_generate(snap_key, snap_stale_notifier);
+                        Ok(!stale)
+                    })
+                );
+                if cf_key_count > 0 {
+                    box_try!(writer.finish());
+                }
+            }
+        }
+
+        if stale {
+            error!("{} build_snapshot meets stale db snap", snap_key);
+            return Err(Error::Stale);
+        }
+
+        // TODO: calculate written bytes and crc32 while scaning.
+        self.written_bytes = get_file_size(&self.tmp_cf_path)?;
+        self.digest = Digest::new_with_initial(crc32::IEEE, calc_crc32(&self.tmp_cf_path)?);
+
+        SNAPSHOT_CF_KV_COUNT
+            .with_label_values(&[self.cf])
+            .observe(cf_key_count as f64);
+        SNAPSHOT_CF_SIZE
+            .with_label_values(&[self.cf])
+            .observe(cf_size as f64);
+        info!(
+            "[region {}] scan snapshot on cf {}, key count {}, size {}",
+            region_id, self.cf, cf_key_count, cf_size
+        );
+        Ok(cf_key_count)
+    }
+}
+
+struct SnapshotBase {
+    dir: String,
+    for_send: bool,
+    key: SnapKey,
+
+    io_limiter: Option<Arc<IOLimiter>>,
+    tmp_cf_files: Vec<CfFile>,
+    success: bool,
+}
+
+impl SnapshotBase {
+    fn new(
+        dir: String,
+        for_send: bool,
+        key: SnapKey,
+        io_limiter: Option<Arc<IOLimiter>>,
+    ) -> SnapshotBase {
+        SnapshotBase {
+            dir,
+            for_send,
+            key,
+            io_limiter,
+            tmp_cf_files: vec![],
+            success: false,
+        }
+    }
+
+    // Rename tmp files to cf files and meta file.
+    // Then collect meta infos into a `SnapshotMeta` and return it.
+    fn save(&mut self) -> Result<SnapshotMeta> {
+        let mut snapshot_meta = SnapshotMeta::new();
+        for cf_builder in &self.tmp_cf_files {
+            let mut snap_cf_file = SnapshotCFFile::new();
+            snap_cf_file.set_cf(cf_builder.cf.to_owned());
+            snap_cf_file.set_size(cf_builder.written_bytes);
+            snap_cf_file.set_checksum(cf_builder.digest.sum32());
+            snapshot_meta.mut_cf_files().push(snap_cf_file);
+        }
+        let tmp_meta_path = gen_meta_tmp_file_path(&self.dir, self.for_send, self.key);
+        let mut tmp_meta_file = create_new_file_at(&tmp_meta_path)?;
+        box_try!(snapshot_meta.write_to_writer(&mut tmp_meta_file));
+        tmp_meta_file.flush()?;
```

Comment:
do we sync the data file?

Reply:
For generating they are just flushed but not synced. Seems it's not very neccesary because we flush and sync after receiving, and check the size and checksum before applying.